crates/storage/snapshot/tree.rs

```diff
@@ -0,0 +1,897 @@
+        // Dive until end or disk layer.
+        let mut diff_wrapped = Layer::DiffLayer(diff.clone());
+        for _ in 0..(layers_n - 1) {
+            match diff_wrapped {
+                Layer::DiskLayer(_) => {
+                    // Diff stack is shallow, no need to modify.
+                    return Ok(None);
+                }
+                Layer::DiffLayer(diff_layer) => {
+                    let diff_value = diff_layer
+                        .read()
+                        .map_err(|error| SnapshotError::LockError(error.to_string()))?;
+                    diff_wrapped = layers[&diff_value.parent()].clone();
+                }
+            }
+        }
+
+        let diff = match diff_wrapped {
+            Layer::DiskLayer(_) => return Ok(None), // should be unreachable
+            Layer::DiffLayer(diff) => diff,
+        };
+
+        let (parent, parent_block_hash) = {
+            let diff_value = diff
+                .read()
+                .map_err(|error| SnapshotError::LockError(error.to_string()))?;
+            (layers[&diff_value.parent()].clone(), diff_value.parent())
+        };
+
+        let parent = match parent {
+            Layer::DiskLayer(_) => return Ok(None),
+            Layer::DiffLayer(diff) => diff,
+        };
+
+        {
+            // hold write lock until linked to new parent to avoid incorrect external reads
+            let mut diff_value = diff
+                .write()
+                .map_err(|error| SnapshotError::LockError(error.to_string()))?;
+
+            // flatten parent into grand parent.
+            let flattened = Self::flatten_diff(parent_block_hash, layers)?;
+            let flattened_block_hash = match &flattened {
+                Layer::DiskLayer(disk_layer) => disk_layer.block_hash(),
+                Layer::DiffLayer(diff_layer) => diff_layer
+                    .read()
+                    .map_err(|error| SnapshotError::LockError(error.to_string()))?
+                    .block_hash(),
+            };
+            layers.insert(flattened_block_hash, flattened.clone());
+            diff_value.set_parent(flattened_block_hash);
+        }
+
+        // Persist the bottom most layer
+        let base = self.save_diff(Layer::DiffLayer(parent), layers)?;
+        layers.insert(base.block_hash, Layer::DiskLayer(base.clone()));
+        let mut diff_value = diff
+            .write()
+            .map_err(|error| SnapshotError::LockError(error.to_string()))?;
+        diff_value.set_parent(base.block_hash());
+        //diff_value.origin = base.block_hash();
+
+        Ok(Some(base))
+    }
+
+    /// Merges the diff into the disk layer.
+    ///
+    /// Returning a new disk layer whose block hash is the diff block hash.
+    ///
+    /// Returns Err if the current disk layer is already marked stale.
+    fn save_diff(
+        &self,
+        diff: Layer,
+        layers: &HashMap<H256, Layer>,
+    ) -> Result<Arc<DiskLayer>, SnapshotError> {
```

Comment:
I feel maybe this should be a disklayer function and just returns the new disklayer or mutates.